# 12 Advanced Recurrent Neural Networks

Advanced Neural Network architectures represent significant advancements in the field of deep learning, which are most used in the domain of sequence modeling and processing. These architectures build upon the traditional feedforward neural networks and introduce recurrent connections, allowing them to exhibit temporal dynamics and memory capabilities.

* The Elman RNN employs a simple recurrent loop in its hidden layer, enabling it to capture short-term temporal dependencies, making it suitable for applications such as speech recognition and time series analysis.
* The Jordan RNN possesses feedback connections from the output layer to the hidden layer, rendering it capable of modeling longer-term dependencies, which finds applications in machine translation and language modeling tasks.
* The Bidirectional RNN combines both forward and backward temporal processing, allowing it to consider both past and future context in its predictions, making it effective in natural language processing tasks such as sentiment analysis and named entity recognition.

These advanced neural network architectures significantly expand the modeling capabilities of traditional neural networks and have become indispensable tools in various sequential data processing applications.

### Exercise
Use the IMDB movie reviews dataset to perform sentiment analysis with a Elman, Jordan and Bidirectional RNN.
Highlight the differences on the performance of each architecture.

In [16]:
from keras.datasets import imdb
#from keras.preprocessing import sequence
from keras.utils.data_utils import pad_sequences
from keras.models import Sequential
from keras.layers import Embedding, SimpleRNN, Dense, Bidirectional
from tensorflow.keras.preprocessing.sequence import pad_sequences

1. Load the IMDB movie reviews dataset

In [17]:
max_features = 5000  # Number of words to consider as features
max_len_short = 100  # Maximum sequence length for short sequences
max_len_long = 500   # Maximum sequence length for long sequences

(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)

2. Pad sequences to a fixed length for RNN input

In [18]:
#!pip install --upgrade keras

In [19]:
x_train_short = pad_sequences(x_train, maxlen=max_len_short)
x_test_short = pad_sequences(x_test, maxlen=max_len_short)

x_train_long = pad_sequences(x_train, maxlen=max_len_long)
x_test_long = pad_sequences(x_test, maxlen=max_len_long)

3. Build the distinct RNN models

In [27]:
def build_elman_rnn_model():
    model = Sequential()
    model.add(Embedding(max_features, 32))
    model.add(SimpleRNN(32, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    return model

def build_jordan_rnn_model():
    model = Sequential()
    model.add(Embedding(max_features, 32))
    model.add(SimpleRNN(32, activation='relu', return_sequences=False))
    model.add(Dense(1, activation='sigmoid'))
    return model

def build_bidirectional_rnn_model():
    model = Sequential()
    model.add(Embedding(max_features, 32))
    model.add(Bidirectional(SimpleRNN(32, activation='relu')))
    model.add(Dense(1, activation='sigmoid'))
    return model

4. Train and evaluate the RNN model

In [21]:
def train_and_evaluate_model(model, x_train, y_train, x_test, y_test):
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    history = model.fit(x_train, y_train, epochs=5, batch_size=128, validation_split=0.2)
    loss, accuracy = model.evaluate(x_test, y_test)
    return loss, accuracy, history

5. Train and evaluate the RNN model on short sequences

In [22]:
print("\nTraining RNN model on short sequences:")
rnn_model_short = build_elman_rnn_model()
loss_short, accuracy_short, history_short = train_and_evaluate_model(
    rnn_model_short, x_train_short, y_train, x_test_short, y_test
)


Training RNN model on short sequences:
Epoch 1/5
157/157 [==============================] - 46s 257ms/step - loss: 0.5860 - accuracy: 0.6706 - val_loss: 0.3944 - val_accuracy: 0.8172
Epoch 2/5
157/157 [==============================] - 26s 164ms/step - loss: 0.3403 - accuracy: 0.8545 - val_loss: 0.3621 - val_accuracy: 0.8362
Epoch 3/5
157/157 [==============================] - 22s 143ms/step - loss: 0.2735 - accuracy: 0.8888 - val_loss: 0.3543 - val_accuracy: 0.8444
Epoch 4/5
157/157 [==============================] - 21s 136ms/step - loss: 0.2328 - accuracy: 0.9070 - val_loss: 0.3975 - val_accuracy: 0.8436
Epoch 5/5
782/782 [==============================] - 10s 13ms/step - loss: 0.3700 - accuracy: 0.8415


6. Train and evaluate the RNN model on long sequences

In [23]:
print("\nTraining Elman RNN model on long sequences:")
rnn_model_long = build_elman_rnn_model()
loss_long, accuracy_long, history_long = train_and_evaluate_model(
    rnn_model_long, x_train_long, y_train, x_test_long, y_test
)


Training Elman RNN model on long sequences:
Epoch 1/5
157/157 [==============================] - 120s 742ms/step - loss: 0.6692 - accuracy: 0.6077 - val_loss: 0.5767 - val_accuracy: 0.7492
Epoch 2/5
157/157 [==============================] - 106s 674ms/step - loss: 0.4402 - accuracy: 0.8090 - val_loss: 0.4010 - val_accuracy: 0.8364
Epoch 3/5
157/157 [==============================] - 103s 654ms/step - loss: 0.3359 - accuracy: 0.8608 - val_loss: 0.3628 - val_accuracy: 0.8468
Epoch 4/5
157/157 [==============================] - 95s 606ms/step - loss: 0.3643 - accuracy: 0.8565 - val_loss: 0.5071 - val_accuracy: 0.7492
Epoch 5/5
782/782 [==============================] - 38s 49ms/step - loss: 0.3395 - accuracy: 0.8638


7. Compare the results

In [28]:
print("\nResults on Short Sequences:")
print(f"Loss: {loss_short:.4f}, Accuracy: {accuracy_short:.4f}")

print("\nResults on Long Sequences:")
print(f"Loss: {loss_long:.4f}, Accuracy: {accuracy_long:.4f}")


Results on Short Sequences:
Loss: 0.3700, Accuracy: 0.8415

Results on Long Sequences:
Loss: 0.3395, Accuracy: 0.8638


In [29]:

print("\nTraining Jordan RNN model on short sequences:")
jordan_rnn_model_short = build_jordan_rnn_model()
loss_short_jordan, accuracy_short_jordan, history_short_jordan = train_and_evaluate_model(
    jordan_rnn_model_short, x_train_short, y_train, x_test_short, y_test
)



Training Jordan RNN model on short sequences:
Epoch 1/5
157/157 [==============================] - 38s 232ms/step - loss: 0.6035 - accuracy: 0.6688 - val_loss: 0.4554 - val_accuracy: 0.7834
Epoch 2/5
157/157 [==============================] - 26s 167ms/step - loss: 0.3758 - accuracy: 0.8339 - val_loss: 0.3614 - val_accuracy: 0.8388
Epoch 3/5
157/157 [==============================] - 20s 128ms/step - loss: 0.2984 - accuracy: 0.8756 - val_loss: 0.3690 - val_accuracy: 0.8386
Epoch 4/5
157/157 [==============================] - 23s 144ms/step - loss: 0.2565 - accuracy: 0.8962 - val_loss: 0.4187 - val_accuracy: 0.8364
Epoch 5/5
782/782 [==============================] - 10s 12ms/step - loss: 0.3751 - accuracy: 0.8391


In [30]:

print("\nTraining Jordan RNN model on long sequences:")
jordan_rnn_model_long = build_jordan_rnn_model()
loss_long_jordan, accuracy_long_jordan, history_long_jordan = train_and_evaluate_model(
    jordan_rnn_model_long, x_train_long, y_train, x_test_long, y_test
)



Training Jordan RNN model on long sequences:
Epoch 1/5
157/157 [==============================] - 111s 697ms/step - loss: 0.6790 - accuracy: 0.5780 - val_loss: 0.6527 - val_accuracy: 0.6322
Epoch 2/5
157/157 [==============================] - 106s 673ms/step - loss: 0.5649 - accuracy: 0.7149 - val_loss: 0.5236 - val_accuracy: 0.7362
Epoch 3/5
157/157 [==============================] - 99s 632ms/step - loss: 0.4080 - accuracy: 0.8268 - val_loss: 0.3917 - val_accuracy: 0.8266
Epoch 4/5
157/157 [==============================] - 99s 633ms/step - loss: 0.2859 - accuracy: 0.8820 - val_loss: 0.3400 - val_accuracy: 0.8508
Epoch 5/5
782/782 [==============================] - 38s 48ms/step - loss: 0.3672 - accuracy: 0.8475


In [33]:
print("\nResults on Short Sequences:")
print(f"Loss: {loss_short_jordan:.4f}, Accuracy: {accuracy_short_jordan:.4f}")

print("\nResults on Long Sequences:")
print(f"Loss: {loss_long_jordan:.4f}, Accuracy: {accuracy_long_jordan:.4f}")


Results on Short Sequences:
Loss: 0.3751, Accuracy: 0.8391

Results on Long Sequences:
Loss: 0.3672, Accuracy: 0.8475


In [31]:

print("\nTraining Bidirectional RNN model on short sequences:")
bidirectional_rnn_model_short = build_bidirectional_rnn_model()
loss_short_bidirectional, accuracy_short_bidirectional, history_short_bidirectional = train_and_evaluate_model(
    bidirectional_rnn_model_short, x_train_short, y_train, x_test_short, y_test
)



Training Bidirectional RNN model on short sequences:
Epoch 1/5
157/157 [==============================] - 51s 301ms/step - loss: 0.6680 - accuracy: 0.5946 - val_loss: 0.5857 - val_accuracy: 0.7112
Epoch 2/5
157/157 [==============================] - 39s 250ms/step - loss: 0.3884 - accuracy: 0.8281 - val_loss: 0.3598 - val_accuracy: 0.8416
Epoch 3/5
157/157 [==============================] - 37s 233ms/step - loss: 0.2768 - accuracy: 0.8884 - val_loss: 0.3552 - val_accuracy: 0.8410
Epoch 4/5
157/157 [==============================] - 38s 240ms/step - loss: 0.2212 - accuracy: 0.9122 - val_loss: 0.4003 - val_accuracy: 0.8264
Epoch 5/5
782/782 [==============================] - 17s 21ms/step - loss: 0.4007 - accuracy: 0.8363


In [32]:

print("\nTraining Bidirectional RNN model on long sequences:")
bidirectional_rnn_model_long = build_bidirectional_rnn_model()
loss_long_bidirectional, accuracy_long_bidirectional, history_long_bidirectional = train_and_evaluate_model(
    bidirectional_rnn_model_long, x_train_long, y_train, x_test_long, y_test
)



Training Bidirectional RNN model on long sequences:
Epoch 1/5
157/157 [==============================] - 189s 1s/step - loss: 0.6113 - accuracy: 0.6618 - val_loss: 0.7538 - val_accuracy: 0.7106
Epoch 2/5
157/157 [==============================] - 178s 1s/step - loss: 0.5190 - accuracy: 0.7667 - val_loss: 0.5211 - val_accuracy: 0.7560
Epoch 3/5
157/157 [==============================] - 166s 1s/step - loss: 0.4400 - accuracy: 0.8115 - val_loss: 0.4428 - val_accuracy: 0.8032
Epoch 4/5
157/157 [==============================] - 160s 1s/step - loss: 0.4030 - accuracy: 0.8275 - val_loss: 0.4095 - val_accuracy: 0.8302
Epoch 5/5
782/782 [==============================] - 73s 93ms/step - loss: 0.3773 - accuracy: 0.8397


In [34]:
print("\nResults on Short Sequences:")
print(f"Loss: {loss_short_bidirectional:.4f}, Accuracy: {accuracy_short_bidirectional:.4f}")

print("\nResults on Long Sequences:")
print(f"Loss: {loss_long_bidirectional:.4f}, Accuracy: {accuracy_long_bidirectional:.4f}")


Results on Short Sequences:
Loss: 0.4007, Accuracy: 0.8363

Results on Long Sequences:
Loss: 0.3773, Accuracy: 0.8397


# RESULTADOS
---
## Elman RNN model
Results on Short Sequences:
Loss: 0.3700, Accuracy: 0.8415

Results on Long Sequences:
Loss: 0.3395, Accuracy: 0.8638

----
## Jordan RNN model
Results on Short Sequences:
Loss: 0.3751, Accuracy: 0.8391

Results on Long Sequences:
Loss: 0.3672, Accuracy: 0.8475

----
## Bidirectional RNN model

Results on Short Sequences:
Loss: 0.4007, Accuracy: 0.8363

Results on Long Sequences:
Loss: 0.3773, Accuracy: 0.8397